group_global_initializers

juj opened the pull request:
Group all global initializers into one function and only export that function to JS scope to save code size.

juj commented:
This resolves #7818. Turns it this was really easy to do in the asm.js module generation side, rather than adding a special LLVM pass. Although this is probably not quite as good as if LLVM would do it, for inlining purposes.

Also this is disabled in EVAL_CTORS=1 builds, since that would prevent ctor evalling. It would be cool to first eval ctors, and then after evalling group the remaining ones together if there are more than one, but that was a bit too hard to do.

juj commented:
I don't think the complexity is that much extra to keep both options around, it is just a check over if EVAL_CTORS is in use or not. This has the advantage that one can test with both forms and see which helps code size more?

kripken commented:
I'd just like to remove all unnecessary complexity in the code, and to remove unnecessary differences between the asm.js and wasm backend code paths. Have you seen a significant benefit right now with EVAL_CTORS, that justifies keeping it around? My general impression was that it needs to be improved anyhow (which I've not gotten around to).

juj commented:
Adjusted the code to centralize the decision to group constructors to a single common function:

```js
def should_group_global_initializers(initializers):
  # If we have at most one global ctor, no need to group global initializers.
  # Also in EVAL_CTORS mode, we want to try to evaluate the individual ctor functions, so in that mode,
  # do not group ctors into one.
  return len(initializers) > 1 and not shared.Settings.EVAL_CTORS
```

There still needs to exist a path that decides whether to group or not, because if there are only one constructor, we do not want to create a dummy indirection, and if there are no ctors, we do want to avoid doing any work. If I changed this to group also when EVAL_CTORS is used, functionally that would only mean changing

```diff
diff --git a/emscripten.py b/emscripten.py
index 3625d5b23..5c813e94d 100644
--- a/emscripten.py
+++ b/emscripten.py
@@ -406,7 +406,7 @@ def should_group_global_initializers(initializers):
   # If we have at most one global ctor, no need to group global initializers.
   # Also in EVAL_CTORS mode, we want to try to evaluate the individual ctor functions, so in that mode,
   # do not group ctors into one.
-  return len(initializers) > 1 and not shared.Settings.EVAL_CTORS
+  return len(initializers) > 1
```

plus wrangling tests, which I'd rather not spend time to go into now, that is not what I am after here. I don't know if ctor evalling gives better or worse code than grouping ctors, and I don't have a test case at hand right now to do comparisons, but I would not mind being able to compare later when I do.

If I make ctor grouping override EVAL_CTORS, then ctor evalling feature will likely go into some kind of dysfunctional mode where it probably will not be able to function at all, so given how easy this is to disable when EVAL_CTORS is enabled, I'd prefer keeping it in for now. (I do not wish to go janitor the test suite right now if I went and disabled EVAL_CTORS from functioning at all with this)

If you'd like to go and drop EVAL_CTORS as a followup cleanup after this PR, I don't oppose. I just don't have data to say if it is functional or not.

juj commented on emscripten.py:
```diff
@@ -776,7 +804,12 @@ def get_exported_implemented_functions(all_exported_functions, all_implemented,
-  funcs = list(funcs) + metadata['initializers']
+  funcs = list(funcs)
+  if should_group_global_initializers(metadata['initializers']):
+    funcs += ['globalCtors']
```
It is not possible to adjust `metadata['initializers']` directly because the code flow first creates the JS side, and only then the asm.js side, meaning that we would need to destroy `metadata['initializers']` already on JS side creation, but still would need to remember what the values of the initializers were in order to create the asm.js side grouped initializer.

Updated the code to avoid duplicate ifs in a slightly different way to get the same effect.
